app/goprotoc/codegen.go

```diff
@@ -38,61 +39,150 @@ var protocOutputs = map[string]struct{}{
+	return filepath.Join(f.path, f.fileName)
+}
+
 func doCodeGen(outputs map[string]string, fds []*desc.FileDescriptor, pluginDefs map[string]string) error {
-	req := plugins.CodeGenRequest{
-		Files:         fds,
-		ProtocVersion: protocVersionStruct,
+	locations, args, err := computeOutputLocations(outputs)
+	if err != nil {
+		return err
 	}
-	resps := map[string]*plugins.CodeGenResponse{}
-	locations := map[string]string{}
+
+	resps, err := runPlugins(args, fds, pluginDefs)
+	if err != nil {
+		return err
+	}
+
+	results, err := assembleFileOutputs(resps, locations)
+	if err != nil {
+		return err
+	}
+
+	// now we can accumulate outputs by archive and emit the
+	// normal files
+	archiveResults := map[outputLocation]map[string]io.Reader{}
+	for file, data := range results {
+		if file.isJar || file.isZip {
+			archiveFiles := archiveResults[file.outputLocation]
+			if archiveFiles == nil {
+				archiveFiles = map[string]io.Reader{}
+				archiveResults[file.outputLocation] = archiveFiles
+			}
+			archiveFiles[file.fileName] = data
+		} else {
+			fileName := filepath.Join(file.path, file.fileName)
+			if err := writeFileResult(fileName, data); err != nil {
+				return err
+			}
+		}
+	}
+
+	// finally: emit any archives
+	for location, files := range archiveResults {
+		if err := writeArchiveResult(location.path, location.isJar, files); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func computeOutputLocations(outputs map[string]string) (map[string]outputLocation, map[string]string, error) {
+	locations := map[string]outputLocation{}
+	args := map[string]string{}
 	for lang, loc := range outputs {
-		resp := plugins.NewCodeGenResponse(lang, nil)
-		resps[lang] = resp
 		locParts := strings.SplitN(loc, ":", 2)
-		var arg string
+		var arg, dest string
 		if len(locParts) > 1 {
 			arg = locParts[0]
-			locations[lang] = locParts[1]
+			dest = locParts[1]
 		} else {
-			locations[lang] = loc
-		}
-		// Protoc expects the base output directories to already exist
-		// so we make the same check here.
-		dirPath := locations[lang]
-		// This should have been validated as part of flag parsing but
-		// this double checks.
-		if dirPath == "" {
-			return fmt.Errorf("%s has no output path", dirPath)
-		}
-		// Protoc has special exceptions for .jar and .zip however.
-		// https://github.com/protocolbuffers/protobuf/blob/6dc9832aab87dfdbb1c7b008f4596184e1fd8105/src/google/protobuf/compiler/command_line_interface.cc#L894
-		if ext := filepath.Ext(dirPath); ext == ".jar" || ext == ".zip" {
-			dirPath = filepath.Dir(dirPath)
-		}
-		fileInfo, err := os.Stat(dirPath)
+			dest = loc
+		}
+		if dest == "" {
+			return nil, nil, fmt.Errorf("%s has empty output path", lang)
+		}
+		dest, err := filepath.Abs(dest)
 		if err != nil {
-			return err
+			return nil, nil, fmt.Errorf("failed to compute absolute path for %s output %s: %v", lang, dest, err)
+		}
+		var isZip, isJar bool
+		if ext := strings.ToLower(filepath.Ext(dest)); ext == ".jar" {
```

Comment:
This is probably better done as a switch statement

Also of note is that `protoc` does not do lowercasing I believe (can double check), and not 100% sure if it makes sense, just a thought

Reply:
Cool - changed to switch.

I think if a user tries to put to a name ending in ".ZIP", they probably wanted a zip archive. That's why I did the lower-case.